monkey/common/di_container.py

```diff
@@ -162,6 +161,15 @@ def _resolve_convention(self, type_: Type[T], name: str) -> T:
                 f"Failed to resolve unregistered convention {convention_identifier}"
             )
 
+    def _resolve_parameter(self, parameter: inspect.Parameter) -> Any:
```

Comment:
I'd say that resolving convention is also resolving a parameter, as dependency is nothing but a parameter. Perhaps it would be more readable to have three methods, `resolve_by_convention` (resolve by parameter name), `resolve_by_type` and `resolve_by_default_value`. They all should follow the same pattern of suppressing the errors until it's resolved in one of three ways. If it's not resolved it should raise an `UnresolvableDependencyError` saying that it failed to resolve the dependency. Right now it will raise an error saying "Failed to resolve unregistered type", but "type" is only one possible way to resolve the dependency. TL;DR **One level of abstraction per function**

Reply:
I also don't really like this. I've pushed a fix.